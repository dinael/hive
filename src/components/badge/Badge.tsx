import { FC } from 'react'
import styles from './Badge.module.scss'

export type BadgeProps = {
  text: string
  length?: number
  spot?: boolean
  alt?: boolean
  className?: string
}

export const Badge: FC<BadgeProps> = ({
  text = '',
  length = 0,
  spot = false,
  alt,
  className = '',
  ...props
}: BadgeProps) => {

  const MAX_LENGTH = length >= 2 ? length : 2

  const textShorten = text.slice(0, MAX_LENGTH)

  const badgeStyles = `
    ${styles['badge-component']}
    ${spot ? styles.spot : ''}
    ${alt ? styles['alt'] : ''}
    ${className || ''}
  `

  const badgeTextStyles = `
    ${styles['badge-text']}
    ${spot ? 'sr-only' : ''}
  `
  const badgeText = (
    <strong
      className={badgeTextStyles}
      aria-hidden="true"
      role="presentation">
      {textShorten}
    </strong>
  )

  return (
    <span className={badgeStyles} {...props}>
      {text ? badgeText : null}
   </span>
  )
}

export default Badge
