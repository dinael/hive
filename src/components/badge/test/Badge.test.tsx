import { describe, test, expect } from 'vitest'
import { render, screen } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import defaultStory, * as stories from "../../../stories/basic/Badge.stories"
import { Badge, BadgeProps } from '../Badge'

//Snapshot
// describe( `[${defaultStory.title} ]`, () => {
//   Object.entries(composeStories).forEach(([story, Component]) => (
//     test(`should render ${story}`, () => {
//       const { container } = render(<Component />)
//       expect(container).toMatchSnapshot()
//     })
//   ))
// })

describe('Badge', () => {
  const defaultProps: BadgeProps = {
    text: '+9999',
    length: 5,
    spot: false,
    className: 'className-test'
  }

  test('renders snapshot', () => {
    const { container } = render(<Badge {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  test('renders with correct text', () => {
    const { container } = render(<Badge {...defaultProps} />)
    const badgeElement = container.querySelector('span')
    expect(badgeElement?.textContent).toBe(defaultProps.text)
    expect(badgeElement?.tagName).toBe('SPAN')
    expect(screen.getByText(defaultProps.text))
    expect(container).toMatchSnapshot()
  })

  test('shortens the text correctly', () => {
    const { container } = render(<Badge {...defaultProps} length={3}/>)
    const badgeElement = container.querySelector('span')
    expect(badgeElement?.textContent).toBe('+99')
    expect(screen.getByText('+99'))
    expect(container).toMatchSnapshot()
  })

  test('applies the spot class when prop is set to true', () => {
    const { container } = render(<Badge {...defaultProps} spot />)
    const badgeElement = container.querySelector('span')
    expect(badgeElement?.className).toContain('spot')
    expect(container).toMatchSnapshot()
  })

  test('applies the spot class when prop is set to true check text a11y', () => {
    const { container } = render(<Badge {...defaultProps} spot />)
    const badgeTextElement = container.querySelector('span > strong')
    expect(badgeTextElement?.textContent).toBe(defaultProps.text)
    expect(badgeTextElement?.className).toContain('sr-only')
    expect(container).toMatchSnapshot()
  })

  test('renders link with custom class', () => {
    const { container } = render(<Badge {...defaultProps} spot />)
    const badgeElement = container.querySelector('span')
    expect(badgeElement?.className).toContain(defaultProps.className)
    expect(container).toMatchSnapshot()
  })
})
