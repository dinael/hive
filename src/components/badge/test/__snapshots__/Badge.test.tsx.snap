// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badge > Badge applies alt classes when alt is true 1`] = `
<div>
  <span
    class="
    badge-component
    
    alt
    className-test
  "
  >
    <strong
      aria-hidden="true"
      class="
    badge-text
    
  "
      role="presentation"
    >
      +9999
    </strong>
  </span>
</div>
`;

exports[`Badge > Badge applies spot classes when prop is set to true 1`] = `
<div>
  <span
    class="
    badge-component
    spot
    
    className-test
  "
  >
    <strong
      aria-hidden="true"
      class="
    badge-text
    sr-only
  "
      role="presentation"
    >
      +9999
    </strong>
  </span>
</div>
`;

exports[`Badge > Badge applies spot classes when prop is set to true check text a11y 1`] = `
<div>
  <span
    class="
    badge-component
    spot
    
    className-test
  "
  >
    <strong
      aria-hidden="true"
      class="
    badge-text
    sr-only
  "
      role="presentation"
    >
      +9999
    </strong>
  </span>
</div>
`;

exports[`Badge > Badge renders link with custom class 1`] = `
<div>
  <span
    class="
    badge-component
    spot
    
    className-test
  "
  >
    <strong
      aria-hidden="true"
      class="
    badge-text
    sr-only
  "
      role="presentation"
    >
      +9999
    </strong>
  </span>
</div>
`;

exports[`Badge > Badge renders snapshot 1`] = `
<div>
  <span
    class="
    badge-component
    
    
    className-test
  "
  >
    <strong
      aria-hidden="true"
      class="
    badge-text
    
  "
      role="presentation"
    >
      +9999
    </strong>
  </span>
</div>
`;

exports[`Badge > Badge renders with correct text 1`] = `
<div>
  <span
    class="
    badge-component
    
    
    className-test
  "
  >
    <strong
      aria-hidden="true"
      class="
    badge-text
    
  "
      role="presentation"
    >
      +9999
    </strong>
  </span>
</div>
`;

exports[`Badge > Badge shortens the text correctly 1`] = `
<div>
  <span
    class="
    badge-component
    
    
    className-test
  "
  >
    <strong
      aria-hidden="true"
      class="
    badge-text
    
  "
      role="presentation"
    >
      +99
    </strong>
  </span>
</div>
`;
